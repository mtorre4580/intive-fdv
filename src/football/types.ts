import { FootballPlayer, FootballPlayerDefault } from './core/football.models';

// Constants
export const FETCH_PLAYERS_LOADING = 'FETCH_PLAYERS_LOADING';
export const FETCHED_PLAYERS = 'FETCHED_PLAYERS';
export const FETCH_PLAYERS_ERR = 'FETCH_PLAYERS_ERR';
export const FILTER_PLAYERS = 'FILTER_PLAYERS';

// Interfaces for Actions (Check types)
interface FetchPlayersLoadingAction {
    type: typeof FETCH_PLAYERS_LOADING;
}

interface FetchedPlayersAction {
    type: typeof FETCHED_PLAYERS;
    players: FootballPlayer[];
}

interface FetchPlayersErrorAction {
    type: typeof FETCH_PLAYERS_ERR;
    err: Error;
}

interface FilterPlayersAction {
    type: typeof FILTER_PLAYERS;
    search: FootballPlayerDefault;
}

// Actions
export type FootballActionTypes = FetchPlayersLoadingAction | FetchedPlayersAction | FetchPlayersErrorAction | FilterPlayersAction;