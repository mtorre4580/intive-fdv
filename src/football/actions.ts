import { FootballService } from './core/football.service';
import { FootballPlayer, FootballPlayerDefault } from './core/football.models';
import { FETCH_PLAYERS_LOADING, FETCHED_PLAYERS, FETCH_PLAYERS_ERR, FILTER_PLAYERS, FootballActionTypes }  from './types';
import { Action } from 'redux';
import { AppState } from '../store';
import { ThunkAction } from 'redux-thunk';

// Service for get all players
const footballService = new FootballService();

export function fetchPlayersLoading() : FootballActionTypes {
  return {
    type: FETCH_PLAYERS_LOADING
  }
}

export function fetchedPlayers(players: FootballPlayer[]) : FootballActionTypes {
  return {
    type: FETCHED_PLAYERS,
    players
  }
}

export function fetchPlayersErr(err: Error) : FootballActionTypes {
  return {
    type: FETCH_PLAYERS_ERR, 
    err
  }
}

export function filterPlayers(search: FootballPlayerDefault) : FootballActionTypes {
  return {
    type: FILTER_PLAYERS, 
    search
  }
}

export const fetchPlayers = () : ThunkAction<void, AppState, null, Action<string>> => async dispatch => {
  dispatch(fetchPlayersLoading());
  try {
    const players : FootballPlayer[] = await footballService.getPlayers();
    dispatch(fetchedPlayers(players));
  } catch (err) {
    dispatch(fetchPlayersErr(err))
  }
}