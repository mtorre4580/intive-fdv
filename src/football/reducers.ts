import { FootballActionTypes, FETCH_PLAYERS_LOADING, FETCHED_PLAYERS, FETCH_PLAYERS_ERR, FILTER_PLAYERS } from './types';
import { FootballState, FootballPlayer } from './core/football.models';

/**
 * Initial state football
 */
const initialState : FootballState = {
    players: [],
    loading: false,
    error: false,
    search: null
}

export function footballReducer(state = initialState, action: FootballActionTypes) : FootballState {
    switch (action.type) {
        case FETCH_PLAYERS_LOADING:
            return {
                ...state,
                loading: true,
                players: [],
                error: false
            }
        case FETCHED_PLAYERS:
            return {
                ...state, 
                players: action.players,
                loading: false,
                error: false
            }
        case FETCH_PLAYERS_ERR:
            return {
                ...state,
                error: action.err,
                loading: false
            }
        case FILTER_PLAYERS:
            const { players } = { ...state };
            const { search } = action;
            return {
                ...state, 
                search,
                players: players.filter((p: FootballPlayer) => p.age === search.age || p.position === search.position || p.name.includes(search.name))
            }
        default:
            return state;
    }
}