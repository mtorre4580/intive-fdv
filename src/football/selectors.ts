import { createSelector } from 'reselect';
import { AppState } from '../store';
import { FootballPlayer } from './core/football.models';

const getSearchSelector = (state: AppState) => state.football.search;
const getPlayersSelector = (state: AppState) => state.football.players;

/**
 * nunca use reselect estoy medio verde mucho angular en almundo, verificar si esta bien implementado #FIXME (Si hay tiempo)
 */
export const getPlayers = createSelector(
    [getPlayersSelector, getSearchSelector], 
    (players : FootballPlayer[], search: any) => {
        if (search) {
            const { age, position, name } = search;
            return players.filter(p => (age === '' || parseInt(p.age) === parseInt(age)) && (position === '' || p.position === position) && (name === '' || p.name === name));
        }
        return players;
});