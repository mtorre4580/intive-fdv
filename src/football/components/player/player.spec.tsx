import React from 'react';
import Player from './player';
import { FootballPlayer } from '../../core/football.models';
import renderer from 'react-test-renderer';

describe('<Player />', () => {
   it('should render snapshot', () => {
    const props : FootballPlayer = {
        name: 'Lionel',
        team: 'Argentina',
        age: '35',
        position: 'Keeper'
    };
    const component = renderer.create(<Player {...props} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
    
});