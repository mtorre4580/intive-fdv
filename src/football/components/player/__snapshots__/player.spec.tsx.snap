// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Player /> should render snapshot 1`] = `
<tr>
  <td>
    Lionel
  </td>
  <td>
    Keeper
  </td>
  <td>
    Argentina
  </td>
  <td>
    35
  </td>
</tr>
`;
